src/mito2/src/region_writer.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use api::v1::{Mutation, Rows, WalEntry};
+use common_query::Output;
+use snafu::ResultExt;
+use store_api::logstore::LogStore;
+use store_api::storage::{RegionId, SequenceNumber};
+use tokio::sync::oneshot::Sender;
+
+use crate::error::{Error, Result, WriteGroupSnafu};
+use crate::memtable::KeyValues;
+use crate::region::version::{VersionControlData, VersionControlRef, VersionRef};
+use crate::wal::{EntryId, WalWriter};
+
+/// Context to keep region metadata and buffer write requests.
+///
+pub(crate) struct RegionWriteCtx {
+    /// Id of region to write.
+    region_id: RegionId,
+    /// Version of the region while creating the context.
+    version: VersionRef,
+    /// Next sequence number to write.
+    ///
+    /// The context assigns a unique sequence number for each row.
+    next_sequence: SequenceNumber,
+    /// Next entry id of WAL to write.
+    next_entry_id: EntryId,
+    /// Valid WAL entry to write.
+    ///
+    /// We keep [WalEntry] instead of mutations to avoid taking mutations
+    /// out of the context to construct the wal entry when we write to the wal.
+    wal_entry: WalEntry,
+    /// Notifiers to send write results to waiters.
+    ///
+    /// The i-th notify is for i-th mutation.
+    notifiers: Vec<WriteNotify>,
+}
+
+impl RegionWriteCtx {
+    /// Returns an empty context.
+    pub(crate) fn new(region_id: RegionId, version_control: VersionControlRef) -> RegionWriteCtx {
```

Comment:
non-blocking: We can pass version_control by ref since we only need to call `current()`.
```suggestion
    pub(crate) fn new(region_id: RegionId, version_control: &VersionControlRef) -> RegionWriteCtx {
```